Library/Homebrew/keg_relocate.rb

```diff
@@ -163,6 +167,45 @@ def replace_text_in_files(relocation, files: nil)
     changed_files
   end
 
+  def replace_build_prefix_in_file(old_cellar, new_prefix, new_cellar, file)
+    old_prefix = Pathname(old_cellar).parent.to_s
+
+    old_prefix_optdir = "#{old_prefix}/opt"
+    old_prefix_etcdir = "#{old_prefix}/etc"
+    old_prefix_vardir = "#{old_prefix}/var"
+    old_prefix_vim_sharedir = "#{old_prefix}/share/vim"
+
+    new_prefix_optdir = "#{new_prefix}/opt"
+    new_prefix_etcdir = "#{new_prefix}/etc"
+    new_prefix_vardir = "#{new_prefix}/var"
+    new_prefix_vim_sharedir = "#{new_prefix}/share/vim"
+
+    old_prefix_dirs = [old_cellar, old_prefix_optdir, old_prefix_etcdir, old_prefix_vardir, old_prefix_vim_sharedir]
+    old_prefix_regex = Regexp.new(old_prefix_dirs.join("|"))
+
+    file.ensure_writable do
+      binary = File.binread file
+      # Skip patching if no string references to the cellar exist.
+      next if binary.scan(old_prefix_regex).empty?
+
+      # Split binary by null characters into array and substitute new cellar for old cellar.
+      # Null padding is added if the new string is too short.
+      patched_binary_strings = []
+      binary.split("\x00").each do |s|
+        patched_binary_strings << s.gsub(old_cellar, new_cellar)
+                                   .gsub(old_prefix_optdir, new_prefix_optdir)
+                                   .gsub(old_prefix_etcdir, new_prefix_etcdir)
+                                   .gsub(old_prefix_vardir, new_prefix_vardir)
+                                   .gsub(old_prefix_vim_sharedir, new_prefix_vim_sharedir)
+                                   .ljust(s.size, "\x00")
```

Comment:
Does this add the padding in the right place? Suppose I had
```
${prefix}/opt${non-null-bytes}${prefix}/var
```
Shouldn't the null bytes be inserted after `${prefix}/opt` instead of at the end of the substring above?

Reply:
In the above example, that string is treated as a single string as far as the ELF/MachO interpreter is concerned.  The goal here is make sure the number of bytes between `${null-byte}${prefix}/opt${non-null-bytes}${prefix}/var${null-byte}` remains the same after prefix substitution.  If the substituted string is shorter than the original, the added null padding at the end effectively makes the string shorter, with the extra null characters being ignored as padding, while keeping the number of bytes the same.  

The ELF/MachO interpreter doesn't know anything about how the string between the null bytes might be processed by program using it.  I don't think we should be changing whatever separator is used for substrings between the null bytes because that is going to be specific to the program.  As it is implemented right now, this should just substitute the new prefix in but leaves everything else exactly as it was, on a per string basis.  The program shouldn't even know that anything was ever changed.